File: src/Aether.Transport/Fec/PolarSCLCodec.cs
// SPDX-License-Identifier: MIT
// Polar erasure codec — Arıkan (2009) butterfly for symbol-level erasure recovery.
//
// Transform (F^T⊗n descending butterfly):
//   for step = N/2 down to 1:
//     for each group block b, offset s in 0..step-1:
//       u[b+s] ^= u[b+s+step]
//
// This produces a lower-triangular generator matrix G where G[r][c] is the
// c-th output of butterfly(e_r).  Each "bit" is a full FixedSymbolSizeBytes
// block; XOR replaces GF(2) addition.
//
// Reliability (Bhattacharyya, BEC p=0.5):
//   z[0] = 0.5;  z[2i] = z[i]^2;  z[2i+1] = 2*z[i] - z[i]^2
//
// Frozen set rule (verified by N=4 MDS check):
//   Freeze nFrozen positions with the SMALLEST z values.
//   Info positions  = K positions with the LARGEST z values.
//
// Decode: GF(2) Gaussian elimination on sub-matrix M[K×K] extracted from G,
//   where M[j][i] = G[infoPos[i]][recvPos[j]], solved simultaneously with
//   the received symbol byte-vectors as the RHS.
//
// Typical use: BLE ≤ 512 B blocks, N=8 coded symbols, K=6 source symbols.

using System;
using System.Collections.Generic;
using Aether.Transport.Abstractions;

namespace Aether.Transport.Fec;

/// <summary>
/// Arıkan polar erasure codec operating on 64-byte symbols.
/// Implements <see cref="IFecCodec"/>; suitable for BLE short-block erasure recovery.
/// Any K-of-N received symbols are sufficient to reconstruct the K source symbols.
/// </summary>
public sealed class PolarSCLCodec : IFecCodec
{
    // ─────────────────────────────────────────────────────────────────
    //  IFecCodec properties
    // ─────────────────────────────────────────────────────────────────

    /// <inheritdoc/>
    public string CodecName => "Polar-SCL";

    /// <inheritdoc/>
    public byte DeviceTierRequired => 1;   // constrained device tier

    /// <inheritdoc/>
    public double OverheadFraction => 0.30;

    /// <inheritdoc/>
    public int FixedSymbolSizeBytes => 64;

    // ─────────────────────────────────────────────────────────────────
    //  Encode
    // ─────────────────────────────────────────────────────────────────

    /// <inheritdoc/>
    /// <remarks>
    /// Packs source bytes into K = ceil(source.Length / 64) info symbols,
    /// zero-pads the last symbol, places them at the info positions of a
    /// length-N polar codeword, applies the butterfly and returns the first
    /// <paramref name="targetSymbolCount"/> coded symbols as a flat byte array.
    /// </remarks>
    public byte[] Encode(ReadOnlySpan<byte> source, int targetSymbolCount)
    {
        int S = FixedSymbolSizeBytes;
        int K = (source.Length + S - 1) / S;         // number of source symbols
        int N = NextPow2(Math.Max(targetSymbolCount, K));

        if (targetSymbolCount < K)
            throw new ArgumentException(
                $"targetSymbolCount ({targetSymbolCount}) must be ≥ source symbol count ({K}).",
                nameof(targetSymbolCount));

        int[]    reliability = ComputeReliabilityOrder(N);   // ascending z
        int      nFrozen     = N - K;
        int[]    infoPos     = reliability[nFrozen..];        // K positions with highest z

        // Allocate polar codeword u[N][S]; frozen positions remain zero.
        byte[][] u = AllocSymbols(N, S);

        for (int i = 0; i < K; i++)
        {
            int srcOffset = i * S;
            int copyLen   = Math.Min(S, source.Length - srcOffset);
            source.Slice(srcOffset, copyLen).CopyTo(u[infoPos[i]]);
            // Remaining bytes already zero — correct zero-padding.
        }

        ButterflyTransform(u);

        byte[] output = new byte[targetSymbolCount * S];
        for (int i = 0; i < targetSymbolCount; i++)
            u[i].CopyTo(output, i * S);

        return output;
    }

    // ─────────────────────────────────────────────────────────────────
    //  TryDecode
    // ─────────────────────────────────────────────────────────────────

    /// <inheritdoc/>
    /// <remarks>
    /// Builds the K×K sub-matrix M of the generator matrix G at the columns
    /// corresponding to received-symbol positions and the rows corresponding to
    /// info positions.  Solves [M | x_received] via GF(2) Gaussian elimination
    /// to recover u[infoPos[i]] for each i.
    /// </remarks>
    public bool TryDecode(ReadOnlyMemory<byte>[] receivedSymbols,
                          int sourceSymbolCount,
                          out byte[]? decoded)
    {
        decoded = null;

        int S = FixedSymbolSizeBytes;
        int K = sourceSymbolCount;
        int N = NextPow2(Math.Max(receivedSymbols.Length, K));

        int[]    reliability = ComputeReliabilityOrder(N);
        int      nFrozen     = N - K;
        int[]    infoPos     = reliability[nFrozen..];

        // Collect received column indices (non-erased, exactly S bytes).
        var recvCols = new List<int>(K);
        for (int j = 0; j < receivedSymbols.Length; j++)
            if (receivedSymbols[j].Length == S)
                recvCols.Add(j);

        if (recvCols.Count < K)
            return false;   // too many erasures

        recvCols = recvCols.GetRange(0, K);   // take exactly K

        // G[r][c] = c-th output of ButterflyBool(e_r), computed once per call.
        // For repeated use the caller should cache a PolarSCLCodec instance and
        // call a pre-built generator; the allocation here is O(N²) bools which
        // is small for the BLE block sizes this codec targets (N ≤ 32).
        bool[][] G = BuildGeneratorMatrix(N);

        // Augmented system: K rows, each row has a K-bit matrix part (M[j][i])
        // and an S-byte RHS (the received symbol).
        bool[][]  M   = new bool[K][];
        byte[][]  rhs = new byte[K][];

        for (int j = 0; j < K; j++)
        {
            M[j]   = new bool[K];
            rhs[j] = new byte[S];
            for (int i = 0; i < K; i++)
                M[j][i] = G[infoPos[i]][recvCols[j]];
            receivedSymbols[recvCols[j]].Span.CopyTo(rhs[j]);
        }

        // GF(2) Gaussian elimination (reduced row echelon).
        for (int col = 0; col < K; col++)
        {
            // Find pivot.
            int pivot = -1;
            for (int row = col; row < K; row++)
            {
                if (M[row][col]) { pivot = row; break; }
            }
            if (pivot < 0)
                return false;   // singular — cannot recover (too many correlated erasures)

            if (pivot != col)
            {
                (M[col], M[pivot])     = (M[pivot], M[col]);
                (rhs[col], rhs[pivot]) = (rhs[pivot], rhs[col]);
            }

            // Eliminate all other rows.
            for (int row = 0; row < K; row++)
            {
                if (row == col || !M[row][col]) continue;
                for (int c = 0; c < K; c++)
                    M[row][c] ^= M[col][c];
                XorInPlace(rhs[row], rhs[col]);
            }
        }

        // rhs[i] now holds u[infoPos[i]].  Concatenate to reconstruct source.
        byte[] result = new byte[K * S];
        for (int i = 0; i < K; i++)
            rhs[i].CopyTo(result, i * S);

        decoded = result;
        return true;
    }

    // ─────────────────────────────────────────────────────────────────
    //  Private helpers — butterfly
    // ─────────────────────────────────────────────────────────────────

    /// <summary>
    /// Applies the F^T⊗n descending butterfly in-place on u[N][S].
    ///
    /// For step = N/2 down to 1, for each block boundary b and offset s in [0, step):
    ///   u[b+s] ^= u[b+s+step]
    ///
    /// This is the "lower-index receives XOR of upper-index" direction that
    /// produces a lower-triangular generator matrix and a verified MDS erasure
    /// code when frozen positions are chosen by ascending Bhattacharyya z-order.
    /// </summary>
    private static void ButterflyTransform(byte[][] u)
    {
        int N = u.Length;
        for (int step = N >> 1; step >= 1; step >>= 1)
            for (int b = 0; b < N; b += step << 1)
                for (int s = 0; s < step; s++)
                    XorInPlace(u[b + s], u[b + s + step]);
    }

    /// <summary>
    /// Boolean butterfly on a single N-element row vector — used to build G.
    /// </summary>
    private static bool[] ButterflyBool(bool[] row)
    {
        bool[] u = (bool[])row.Clone();
        int N = u.Length;
        for (int step = N >> 1; step >= 1; step >>= 1)
            for (int b = 0; b < N; b += step << 1)
                for (int s = 0; s < step; s++)
                    u[b + s] ^= u[b + s + step];
        return u;
    }

    // ─────────────────────────────────────────────────────────────────
    //  Private helpers — generator matrix & reliability
    // ─────────────────────────────────────────────────────────────────

    /// <summary>
    /// Builds the N×N generator matrix G where G[r] = ButterflyBool(e_r).
    /// G[r][c] is true iff coded symbol c involves input symbol r.
    /// </summary>
    private static bool[][] BuildGeneratorMatrix(int N)
    {
        bool[][] G = new bool[N][];
        for (int r = 0; r < N; r++)
        {
            bool[] e = new bool[N];
            e[r] = true;
            G[r] = ButterflyBool(e);
        }
        return G;
    }

    /// <summary>
    /// Returns indices 0..N-1 sorted in ascending Bhattacharyya z-order.
    ///
    /// The first nFrozen entries (lowest z) are the frozen positions.
    /// The last K entries (highest z) are the information positions.
    ///
    /// Recursion (BEC with erasure probability p = 0.5):
    ///   z[0]      = 0.5
    ///   z[2i]     = z[i]²             (upper sub-channel — more reliable)
    ///   z[2i+1]   = 2·z[i] − z[i]²   (lower sub-channel — less reliable)
    /// </summary>
    private static int[] ComputeReliabilityOrder(int N)
    {
        double[] z = ComputeBhattacharyya(N);
        int[] order = new int[N];
        for (int i = 0; i < N; i++) order[i] = i;
        Array.Sort(order, (a, b) => z[a].CompareTo(z[b]));
        return order;
    }

    /// <summary>
    /// Computes Bhattacharyya z-values for all N synthetic channels of BEC(0.5).
    /// z[i] is the probability that synthetic channel i is erased.
    /// </summary>
    private static double[] ComputeBhattacharyya(int N)
    {
        double[] z = new double[N];
        z[0] = 0.5;
        for (int step = 1; step < N; step <<= 1)
        {
            // Expand in reverse to avoid overwriting values still needed.
            for (int i = step - 1; i >= 0; i--)
            {
                double zi      = z[i];
                z[2 * i]       = zi * zi;            // upper — better channel
                z[2 * i + 1]   = 2.0 * zi - zi * zi; // lower — worse channel
            }
        }
        return z;
    }

    // ─────────────────────────────────────────────────────────────────
    //  Private helpers — memory
    // ─────────────────────────────────────────────────────────────────

    private static byte[][] AllocSymbols(int count, int symSize)
    {
        byte[][] arr = new byte[count][];
        for (int i = 0; i < count; i++)
            arr[i] = new byte[symSize];
        return arr;
    }

    private static void XorInPlace(byte[] dst, byte[] src)
    {
        for (int i = 0; i < dst.Length; i++)
            dst[i] ^= src[i];
    }

    private static int NextPow2(int n)
    {
        if (n <= 1) return 1;
        int p = 1;
        while (p < n) p <<= 1;
        return p;
    }
}

File: kotlin/src/main/kotlin/aethermesh/transport/fec/PolarSCLCodec.kt
// SPDX-License-Identifier: MIT
// Polar erasure codec — Arıkan (2009) butterfly for symbol-level erasure recovery.
//
// Transform (F^T⊗n descending butterfly):
//   for step = N/2 down to 1:
//     for each group block b, offset s in 0..<step:
//       u[b+s] ^= u[b+s+step]
//
// Produces a lower-triangular generator matrix G; each "bit" is a full
// SYMBOL_BYTES block and XOR replaces GF(2) addition.
//
// Reliability (Bhattacharyya, BEC p = 0.5):
//   z[0] = 0.5;  z[2i] = z[i]²;  z[2i+1] = 2·z[i] − z[i]²
//
// Frozen set rule (verified by N=4 MDS check):
//   Freeze nFrozen positions with the SMALLEST z values.
//   Info positions  = K positions with the LARGEST z values.
//
// Decode: GF(2) Gaussian elimination on sub-matrix M[K×K] extracted from G,
//   where M[j][i] = G[infoPos[i]][recvPos[j]], solved simultaneously with
//   the received symbol byte-vectors as the RHS.

package aethermesh.transport.fec

import aethermesh.transport.FecCodec

/**
 * Arıkan polar erasure codec operating on 64-byte symbols.
 *
 * Implements [FecCodec]; suitable for BLE short-block erasure recovery.
 * Any K-of-N received symbols are sufficient to reconstruct the K source symbols.
 */
class PolarSCLCodec : FecCodec {

    // ─────────────────────────────────────────────────────────────────
    //  FecCodec properties
    // ─────────────────────────────────────────────────────────────────

    override val codecName: String          get() = "Polar-SCL"
    override val deviceTierRequired: Byte   get() = 1
    override val overheadFraction: Double   get() = 0.30
    override val fixedSymbolSizeBytes: Int  get() = SYMBOL_BYTES

    // ─────────────────────────────────────────────────────────────────
    //  Encode
    // ─────────────────────────────────────────────────────────────────

    /**
     * Encodes [source] into [targetSymbolCount] coded symbols.
     *
     * The source is split into K = ceil(source.size / 64) info symbols (last
     * one zero-padded), placed at the K info positions of a length-N polar
     * codeword, the butterfly is applied, and the first [targetSymbolCount]
     * coded symbols are returned as a flat byte array.
     */
    override fun encode(source: ByteArray, targetSymbolCount: Int): ByteArray {
        val S = SYMBOL_BYTES
        val K = (source.size + S - 1) / S
        val N = nextPow2(maxOf(targetSymbolCount, K))

        require(targetSymbolCount >= K) {
            "targetSymbolCount ($targetSymbolCount) must be ≥ source symbol count ($K)"
        }

        val reliability = computeReliabilityOrder(N)   // ascending z
        val nFrozen     = N - K
        val infoPos     = reliability.sliceArray(nFrozen until N)   // K positions with highest z

        // Allocate polar codeword u[N][S]; frozen positions remain zero.
        val u = Array(N) { ByteArray(S) }

        for (i in 0 until K) {
            val srcOffset = i * S
            val copyLen   = minOf(S, source.size - srcOffset)
            source.copyInto(u[infoPos[i]], destinationOffset = 0,
                            startIndex = srcOffset, endIndex = srcOffset + copyLen)
            // Remaining bytes already zero — correct zero-padding.
        }

        butterflyTransform(u)

        val output = ByteArray(targetSymbolCount * S)
        for (i in 0 until targetSymbolCount)
            u[i].copyInto(output, destinationOffset = i * S)

        return output
    }

    // ─────────────────────────────────────────────────────────────────
    //  TryDecode
    // ─────────────────────────────────────────────────────────────────

    /**
     * Attempts to reconstruct the original data from [receivedSymbols].
     *
     * [receivedSymbols] is indexed by coded-symbol position; any element whose
     * [ByteArray.size] ≠ 64 is treated as erased.
     *
     * Returns the reconstructed source bytes if decoding succeeded, or `null`
     * if too many symbols were erased.
     */
    override fun tryDecode(receivedSymbols: List<ByteArray>, sourceSymbolCount: Int): ByteArray? {
        val S = SYMBOL_BYTES
        val K = sourceSymbolCount
        val N = nextPow2(maxOf(receivedSymbols.size, K))

        val reliability = computeReliabilityOrder(N)
        val nFrozen     = N - K
        val infoPos     = reliability.sliceArray(nFrozen until N)

        // Collect received column indices (non-erased, exactly S bytes).
        val recvCols = mutableListOf<Int>()
        for (j in receivedSymbols.indices)
            if (receivedSymbols[j].size == S)
                recvCols += j

        if (recvCols.size < K) return null   // too many erasures

        val chosen = recvCols.subList(0, K)   // take exactly K

        val G = buildGeneratorMatrix(N)

        // Augmented system: K rows × (K-bit matrix + S-byte RHS).
        val M   = Array(K) { j -> BooleanArray(K) { i -> G[infoPos[i]][chosen[j]] } }
        val rhs = Array(K) { j -> receivedSymbols[chosen[j]].copyOf() }

        // GF(2) Gaussian elimination (reduced row echelon).
        for (col in 0 until K) {
            // Find pivot.
            val pivot = (col until K).firstOrNull { M[it][col] } ?: return null

            if (pivot != col) {
                val tmpM = M[col];   M[col] = M[pivot];   M[pivot] = tmpM
                val tmpR = rhs[col]; rhs[col] = rhs[pivot]; rhs[pivot] = tmpR
            }

            for (row in 0 until K) {
                if (row == col || !M[row][col]) continue
                for (c in 0 until K) M[row][c] = M[row][c] xor M[col][c]
                xorInPlace(rhs[row], rhs[col])
            }
        }

        // rhs[i] now holds u[infoPos[i]].  Concatenate to reconstruct source.
        val result = ByteArray(K * S)
        for (i in 0 until K) rhs[i].copyInto(result, destinationOffset = i * S)
        return result
    }

    // ─────────────────────────────────────────────────────────────────
    //  Private helpers — butterfly
    // ─────────────────────────────────────────────────────────────────

    /**
     * Applies the F^T⊗n descending butterfly in-place on u[N][S].
     *
     * For step = N/2 down to 1, for each block boundary b and offset s in [0, step):
     *   u[b+s] ^= u[b+s+step]
     *
     * Lower-index receives XOR of upper-index. This direction produces a
     * lower-triangular G and verified-MDS erasure code for frozen-smallest-z.
     */
    private fun butterflyTransform(u: Array<ByteArray>) {
        val N = u.size
        var step = N shr 1
        while (step >= 1) {
            var b = 0
            while (b < N) {
                for (s in 0 until step)
                    xorInPlace(u[b + s], u[b + s + step])
                b += step shl 1
            }
            step = step shr 1
        }
    }

    /**
     * Boolean butterfly on a single N-element row vector — used to build G.
     */
    private fun butterflyBool(row: BooleanArray): BooleanArray {
        val u = row.copyOf()
        val N = u.size
        var step = N shr 1
        while (step >= 1) {
            var b = 0
            while (b < N) {
                for (s in 0 until step)
                    u[b + s] = u[b + s] xor u[b + s + step]
                b += step shl 1
            }
            step = step shr 1
        }
        return u
    }

    // ─────────────────────────────────────────────────────────────────
    //  Private helpers — generator matrix & reliability
    // ─────────────────────────────────────────────────────────────────

    /**
     * Builds the N×N generator matrix G where G[r] = butterflyBool(e_r).
     * G[r][c] is true iff coded symbol c involves input symbol r.
     */
    private fun buildGeneratorMatrix(N: Int): Array<BooleanArray> =
        Array(N) { r ->
            val e = BooleanArray(N) { it == r }
            butterflyBool(e)
        }

    /**
     * Returns indices 0..(N-1) sorted by ascending Bhattacharyya z-value.
     *
     * First nFrozen entries (lowest z) → frozen positions.
     * Last K entries (highest z)       → information positions.
     *
     * Recursion (BEC, erasure probability p = 0.5):
     *   z[0]    = 0.5
     *   z[2i]   = z[i]²              (upper sub-channel — more reliable)
     *   z[2i+1] = 2·z[i] − z[i]²    (lower sub-channel — less reliable)
     */
    private fun computeReliabilityOrder(N: Int): IntArray {
        val z = computeBhattacharyya(N)
        // Sort indices 0..(N-1) by ascending z-value:
        //   smallest z → frozen (start of array)
        //   largest z  → info   (end of array)
        return IntArray(N) { it }.sortedBy { z[it] }.toIntArray()
    }

    /**
     * Computes Bhattacharyya z-values for all N synthetic channels of BEC(0.5).
     */
    private fun computeBhattacharyya(N: Int): DoubleArray {
        val z = DoubleArray(N)
        z[0] = 0.5
        var step = 1
        while (step < N) {
            // Expand in reverse to avoid overwriting values still needed.
            for (i in step - 1 downTo 0) {
                val zi      = z[i]
                z[2 * i]     = zi * zi
                z[2 * i + 1] = 2.0 * zi - zi * zi
            }
            step = step shl 1
        }
        return z
    }

    // ─────────────────────────────────────────────────────────────────
    //  Private helpers — memory / math
    // ─────────────────────────────────────────────────────────────────

    private fun xorInPlace(dst: ByteArray, src: ByteArray) {
        for (i in dst.indices) dst[i] = (dst[i].toInt() xor src[i].toInt()).toByte()
    }

    private fun nextPow2(n: Int): Int {
        if (n <= 1) return 1
        var p = 1
        while (p < n) p = p shl 1
        return p
    }

    companion object {
        private const val SYMBOL_BYTES = 64
    }
}
